pkg/cmd/roachtest/tests/cluster_to_cluster.go

```diff
@@ -726,10 +755,282 @@ func registerClusterToCluster(r registry.Registry) {
+	case steadyState:
+		return "Steady State"
+	case phaseCutover:
+		return "Cutover"
+	default:
+		return "unknown"
+	}
+}
+
+type replResilienceSpec struct {
+	onSrc         bool
+	onCoordinator bool
+	// phase indicates the c2c phase a node shutdown will occur.
+	phase c2cPhase
+	spec  *replicationTestSpec
+
+	// the fields below are gathered after the replication stream has started
+	srcJobID     jobspb.JobID
+	dstJobID     jobspb.JobID
+	shutdownNode int
+	watcherNode  int
+}
+
+func (rsp *replResilienceSpec) name() string {
+	var builder strings.Builder
+	builder.WriteString("c2c/shutdown")
+	if rsp.onSrc {
+		builder.WriteString("/src")
+	} else {
+		builder.WriteString("/dest")
+	}
+	if rsp.onCoordinator {
+		builder.WriteString("/coordinator")
+	} else {
+		builder.WriteString("/worker")
+	}
+	return builder.String()
+}
+func (rsp *replResilienceSpec) getJobIDs(ctx context.Context) {
+	setup := rsp.spec.setup
+	jobIDQuery := `SELECT job_id FROM [SHOW JOBS] WHERE job_type = '%s'`
+	testutils.SucceedsWithin(rsp.spec.t, func() error {
+		if err := setup.dst.db.QueryRowContext(ctx, fmt.Sprintf(jobIDQuery,
+			`STREAM INGESTION`)).Scan(&rsp.dstJobID); err != nil {
+			return err
+		}
+		if err := setup.src.db.QueryRowContext(ctx, fmt.Sprintf(jobIDQuery,
+			`STREAM REPLICATION`)).Scan(&rsp.srcJobID); err != nil {
+			return err
+		}
+		return nil
+	}, time.Minute)
+}
+
+func (rsp *replResilienceSpec) getTargetInfo() (*clusterInfo, jobspb.JobID, option.NodeListOption) {
+	if rsp.onSrc {
+		return rsp.spec.setup.src, rsp.srcJobID, rsp.spec.setup.src.nodes
+	}
+	return rsp.spec.setup.dst, rsp.dstJobID, rsp.spec.setup.dst.nodes
+}
+
+func (rsp *replResilienceSpec) getTargetAndWatcherNodes(ctx context.Context) {
+	var coordinatorNode int
+	info, jobID, nodes := rsp.getTargetInfo()
+
+	// To populate the coordinator_id field, a node needs to claim the job.
+	// Give the job system a minute.
+	testutils.SucceedsWithin(rsp.spec.t, func() error {
+		return info.db.QueryRowContext(ctx,
+			`SELECT coordinator_id FROM crdb_internal.jobs WHERE job_id = $1`, jobID).Scan(&coordinatorNode)
+	}, time.Minute)
+
+	var targetNode int
+
+	findAnotherNode := func(notThisNode int) int {
+		for {
+			anotherNode := nodes.RandNode()[0]
+			if notThisNode != anotherNode {
+				return anotherNode
+			}
+		}
+	}
+	if rsp.onCoordinator {
+		targetNode = coordinatorNode
+	} else {
+		targetNode = findAnotherNode(coordinatorNode)
+	}
+	rsp.shutdownNode = targetNode
+	rsp.watcherNode = findAnotherNode(targetNode)
+}
+
+func (rsp *replResilienceSpec) getPhase() c2cPhase {
+	progress := getJobProgress(rsp.spec.t, rsp.spec.setup.dst.sysSQL, rsp.dstJobID)
+	streamIngestProgress := progress.GetStreamIngest()
+	highWater := progress.GetHighWater()
+
+	if highWater == nil || highWater.IsEmpty() {
+		return phaseInitialScan
+	}
+	if streamIngestProgress.CutoverTime.IsEmpty() {
+		return steadyState
+	}
+	// TODO check that job has not complete
+	return phaseCutover
```

Comment:
sadly the replication status options currently does not provide a `CutoverDone` option. When i allow this test to shutdown during cutover, I'll plan to resolve the todo by checking the job status instead. Ok with you?

Reply:
ah right.. there is no job when cutover is done, that's ok.
but you can still use that status to figure out whether you're in initial scan or steady state, instead of looking at the cutover time and the high water.

Reply:
yeah, given that ci is green, i'm going bors it and reconsider in the follow up pr :D